KZOO-271: Add optional mechanism to limit concurrent calls per endpoint

harenson opened the pull request:
Depends on: https://github.com/2600hz/kazoo-crossbar/pull/362

This new mechanism allows to set a maximum concurrent calls limit per endpoint (device/user) or account-wide. When configure at device, user and account level, device's configuration takes precedence over user's configuration, and  user's configuration takes precedence over account's configuration.

harenson committed:
KZOO-271: Add optional mechanism to limit concurrent calls per endpoint

This new mechanism allows to set a maximum concurrent calls limit per
endpoint (device/user) or account-wide. When configure at device, user
and account level, device's configuration takes precedence over user's
configuration, and  user's configuration takes precedence over account's
configuration.

harenson committed:
KZOO-271: Document a couple of functions within kz_json module

lazedo commented on kazoo_transactions/src/kz_endpoint_transactions.erl:
```diff
@@ -0,0 +1,64 @@
+-spec release(j5_endpoint_limit_ra:uuid()) -> j5_endpoint_limit_ra:release_resp().
+release(UUID) ->
+    j5_endpoint_limit_ra:release(UUID).
+
```
core libraries shuold not refer to applications

harenson replied:
The endpoint_limit API was moved out of jonny5 and put into core/kazoo_call application instead.

harenson committed:
KZOO-271: Fix copyright tag

harenson committed:
KZOO-271: Move endpoint_limit API into core/kazoo_call app

harenson committed:
KZOO-271: Use lowest call_limits value when building endpoints

When building the endpoint, get the lowest call_limits -> max_concurrent
value from the Account, the Endpoint, and the Owner (if endpoint=device
and device is assigned) and assign it to the endpoint being built.

harenson committed:
KZOO-271: Bump copyright year

harenson committed:
KZOO-271: Abstract examples variables

harenson committed:
KZOO-271: Rename RA cluster name for endpoint_limit

harenson committed:
KZOO-271: Add merge_attribute clause for call_limits setting

When building the endpoint, if limit is configured at endpoint level, it
takes precedence over account's configuration in case limit was also
configured account-wide.

harenson committed:
KZOO-271: Add API to know if endpoint has reached its limit

Builds AMQP request object to query whether or not the given endpoint_id
reached the given limit. Sends the AMQP request object, waits for the
response, reads the response and replies with a boolean based on the
response received.

harenson committed:
KZOO-271: Fix indentation and function doc

harenson committed:
KZOO-271: Add channels per owner-id listing

harenson committed:
KZOO-271: Improve code comments and definitions

harenson committed:
KZOO-271: Improve validation function

danielfinke commented on kazoo_amqp/src/api/kapi_call.erl:
```diff
@@ -426,25 +504,29 @@ query_user_channels_req_definition() ->
               ,{fun kapi_definition:set_types/2
                ,[{<<"Active-Only">>, fun kz_term:is_boolean/1}
                 ,{<<"Authorizing-IDs">>, fun erlang:is_list/1}
+                ,{<<"Owner-ID">>, fun erlang:is_binary/1}
```
should the type for `<<"Realm">>` also be added?

lazedo replied:
don't add this here. you can add a `query_owner_channels` and hangdle it on ecall_fs_channels
also, as a requirement, add a `query_endpoint_channels`  and handle it on ecallmgr_channels with selector as ( authorizing_id = $1 or owner_id = $1 ) and an optional direction with `both` as default and adapt the selector for that

harenson replied:
@lazedo Do you still want me to take care of these changes you requested here?

lazedo replied:
@harenson no, they're already done

danielfinke commented on kazoo_amqp/src/api/kapi_call.erl:
```diff
@@ -718,6 +804,61 @@ publish_channel_status_resp(RespQ, Resp, ContentType) ->
 
+%%------------------------------------------------------------------------------
+%% @doc Sends request to know whether or not an endpoint has reached its configured limit.
+%% Takes {@link kz_term:proplist()}, creates JSON string or error.
```
can maybe omit this line since it's accounted for in the spec
```suggestion
```
same suggestion on L838

harenson replied:
Done.

danielfinke commented on kazoo_amqp/src/api/kapi_call.erl:
```diff
@@ -1113,6 +1263,11 @@ realm(Props) when is_list(Props) ->
     kz_json:get_ne_binary_value(<<"Realm">>, JObj).
 
+%% Routing keys builders
+-spec event_routing_key(kz_term:ne_binary(), kz_term:ne_binary()) -> kz_term:ne_binary().
+event_routing_key(EventName, CallId) ->
+    call_routing_key(EventName, CallId).
+
```
it looks like this fun could be removed if you switch the usage in `kapi_dialplan` to use `call_routing_key/3` instead

harenson replied:
They are separated kapi modules. This routing key creation is per kapi module.

danielfinke commented on kazoo_call/doc/endpoint_limit.md:
```diff
@@ -0,0 +1,56 @@
+false = kapps_endpoint_limit:has_reached_limit(EndpointId, EndpointType, EndpointLimit),
+false = kapps_endpoint_limit:has_reached_limit(OwnerId, <<"user">>, OwnerLimit),
+
+%% If a call is placed from/to EndpointId, check should return true when quering for endpoint's limit
```
typo
```suggestion
%% If a call is placed from/to EndpointId, check should return true when querying for endpoint's limit
```

harenson replied:
Fixed.

danielfinke commented on kazoo_call/src/kapps_call.hrl:
```diff
@@ -0,0 +1,10 @@
+-ifndef(KAPPS_CALL_HEADERS).
```
convention
```suggestion
-ifndef(KAPPS_CALL_HRL).
```

harenson replied:
Changed.

danielfinke commented on kazoo_stdlib/src/kz_maps.erl:
```diff
@@ -18,6 +18,7 @@
 -export([put/3, put/4]).
 -export([keys_to_atoms/1, keys_to_atoms/2]).
 -export([exec/2]).
+-export([remove/2]).
```
should these `kz_maps` changes be split out into a separate PR? They don't seem related

harenson replied:
Removed.

danielfinke commented on kazoo_stdlib/src/kz_maps.erl:
```diff
@@ -168,6 +169,14 @@ exec_fold({Fun, Arg1, Arg2}, Map)
     Fun(Map, Arg1, Arg2).
 
+-spec remove(term() | [term()], map()) -> map().
+remove([Key | Rem], Map) ->
```
this is very similar to `maps:without/2`

danielfinke commented on kazoo_call/src/kapps_endpoint_limit.erl:
```diff
@@ -0,0 +1,63 @@
+%%%-----------------------------------------------------------------------------
+%%% @copyright (C) 2022, 2600Hz
+%%% @doc RA (Raft) integration
```
is this still going to be a part of some Raft setup?

harenson replied:
No, it is not, removing it.

danielfinke reviewed (changes requested):
Looks good to me. Just have some minor asks/questions

lazedo reviewed (changes requested):
marking this as not mergeable

lazedo commented on kazoo_amqp/src/api/kapi_call.erl:
```diff
@@ -32,6 +32,16 @@
         ,publish_channel_status_resp/3
         ,channel_status_extended_headers/0
         ]).
+-export([enforce_endpoint_limit_req/1
+        ,enforce_endpoint_limit_req_v/1
+        ,publish_enforce_endpoint_limit_req/1
+        ,publish_enforce_endpoint_limit_req/2
+        ]).
+-export([enforce_endpoint_limit_resp/1
+        ,enforce_endpoint_limit_resp_v/1
+        ,publish_enforce_endpoint_limit_resp/2
+        ,publish_enforce_endpoint_limit_resp/3
+        ]).
```
this should go into a new kapi, lets leave this one for queries

harenson replied:
@lazedo Do you still want me to move this functions are out of this module and create a new one for it?

lazedo commented on kazoo_stdlib/src/kz_maps.erl:
```diff
@@ -168,6 +169,14 @@ exec_fold({Fun, Arg1, Arg2}, Map)
 
+-spec remove(term() | [term()], map()) -> map().
+remove([Key | Rem], Map) ->
+    remove(Rem, maps:remove(Key, Map));
```
this looks wrong, what's the intent ?
#{a => #{a1 => #{a2 => #{a3 => #{xxx}}}}}
remove([a,a2,a3,xxx]) will remove remove `a` and should only remove `xxxx`

lazedo replied:
if the intent was to remove a list of keys, then as @danielfinke mentioned, `maps:without/2` is to be used and it shouldn't be here

harenson replied:
This was a left-over from previous implementation. Removed.

lazedo reviewed (changes requested):
lets split the pr.

lets merge the changes for the accessors, kapi_call (query_channels_by_owner_id, query_channels_by_endpoint), kazoo_endpoint, kazoo_directory

then lets do the corresponding changes on ecallmgr fs channels

then lets add the core kazoo_call and kapi limits changes 

then lets talk about where to put the listener that will call the channels_limits

also, lets try not to merge unrelated changes

harenson committed:
KZOO-271: Take care of reviewer comments

harenson committed:
KZOO-271: Remove left-over/unneeded function

lazedo commented:
@harenson move this to draft

harenson commented:
@lazedo Done, what changes do you want me to do before marking it as ready for review again?

danielfinke commented on kazoo_call/src/kapps_endpoint_limit.erl:
```diff
@@ -0,0 +1,62 @@
+%%%-----------------------------------------------------------------------------
+%%% @copyright (C) 2022, 2600Hz
+%%% This Source Code Form is subject to the terms of the Mozilla Public
```
missing `@doc`
```suggestion
%%% @doc This Source Code Form is subject to the terms of the Mozilla Public
```

harenson replied:
That is not documentation, but license information. I added a very short documentation to this module, anyway.

danielfinke commented on kazoo_call/src/kapps_call.hrl:
```diff
@@ -0,0 +1,10 @@
+-define(KAPPS_CALL_APP_NAME, <<"kapps_call">>).
+-define(KAPPS_CALL_APP_VERSION, <<"1.0.0">>).
+
+-define(KAPPS_CALL_HEADERS, 'true').
```
need to update this line too
```suggestion
-define(KAPPS_CALL_HRL, 'true').
```

harenson replied:
Nice catch, thank you.

harenson committed:
KZOO-271: Fix macro name and add a short module documentation
